### PR TITLE
Loki: Implement custom /config handler

### PR DESCRIPTION
**Pull request by DylanGuedes**
**What this PR does / why we need it**:
- Creates a new `RunOpts` used by Loki internally to customize running behavior. It is used in this PR to customize how to handle `/config` endpoint and can be used later to customize other behaviors.

**Which issue(s) this PR fixes**:
N/A

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [ ] Tests updated
- [ ] Add an entry in the `CHANGELOG.md` about the changes.

### CODE REVIEW

File: pkg/loki/loki.go [hunk 1]
```diff
@@ -280,8 +280,23 @@ func newDefaultConfig() *Config {
 	return defaultConfig
 }
 
+// RunOpts configures custom behavior for running Loki.
+type RunOpts struct {
+	// customConfigEndpointHandlerFn is the handlerFunc to be used by the /config endpoint.
+	// If empty, default handlerFunc will be used.
+	customConfigEndpointHandlerFn func(http.ResponseWriter, *http.Request)
```
Comment (sandeepsukhani) [inline pkg/loki/loki.go:287]: Should we export it to allow defining custom values? Without that, I do not see a way to change the behaviour outside of `loki` package.

Comment (DylanGuedes) [reply pkg/loki/loki.go:287]: Oops, good point. I'll make it public then.

File: pkg/loki/loki.go [hunk 2]
```diff
@@ -280,8 +280,23 @@ func newDefaultConfig() *Config {
 	return defaultConfig
 }
 
+// RunOpts configures custom behavior for running Loki.
+type RunOpts struct {
+	// CustomConfigEndpointHandlerFn is the handlerFunc to be used by the /config endpoint.
+	// If empty, default handlerFunc will be used.
+	CustomConfigEndpointHandlerFn func(http.ResponseWriter, *http.Request)
```
Comment (jeschkies) [inline pkg/loki/loki.go:287]: Out of curiosity, why is this required?

Comment (DylanGuedes) [reply pkg/loki/loki.go:287]: Without this, other projects (ex: GEL) aren't able to customize the `/config` endpoint. That's a problem for projects that uses a superset of configurations that include Loki configs because their configs that aren't part of Loki aren't shown in the /config endpoint.

Comment (jeschkies) [reply pkg/loki/loki.go:287]: That's a very good point. It relates to https://github.com/grafana/loki-private/issues/96 though :slightly_smiling_face:

Comment (DylanGuedes) [reply pkg/loki/loki.go:287]: Yes, it has some relation to it.

File: pkg/loki/loki.go [hunk 3]
```diff
@@ -280,8 +280,23 @@ func newDefaultConfig() *Config {
...
+}
+
+func (t *Loki) bindConfigEndpoint(opts RunOpts) {
```
Comment (cyriltovena) [inline pkg/loki/loki.go:290]: @MichelHollands Sounds like something you could have used.

**Review by cyriltovena — APPROVED** [review]
LGTM
